tracing/ops_tracing/_api.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""The tracing API for the charms."""
+
+from __future__ import annotations
+
+import logging
+
+import ops
+
+from ._buffer import Destination
+from .vendor.charms.certificate_transfer_interface.v1.certificate_transfer import (
+    CertificateTransferRequires,
+)
+from .vendor.charms.tempo_coordinator_k8s.v0.tracing import (
+    AmbiguousRelationUsageError,
+    ProtocolNotRequestedError,
+    TracingEndpointRequirer,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class Tracing(ops.Object):
+    """Initialise the tracing service.
+
+    Usage:
+        - Include ``ops[tracing]`` in your dependencies.
+        - Declare the relations that the charm supports.
+        - Initialise ``Tracing`` with the names of these relations.
+
+    Example::
+
+        # charmcraft.yaml
+        requires:
+            charm-tracing:
+                interface: tracing
+                limit: 1
+                optional: true
+            send-ca-cert:
+                interface: certificate_transfer
+                limit: 1
+                optional: true
+
+        # src/charm.py
+        import ops.tracing
+
+        class SomeCharm(ops.CharmBase):
+            def __init__(self, framework: ops.Framework):
+                ...
+                self.tracing = ops.tracing.Tracing(
+                    self,
+                    tracing_relation_name="charm-tracing",
+                    ca_relation_name="send-ca-cert",
+                )
+
+    Args:
+        charm: your charm instance
+        tracing_relation_name: the name of the relation that provides the
+            destination to send trace data to.
+        ca_relation_name: the name of the relation that provides the CA
+            list to validate the tracing destination against.
+        ca_data: a fixed CA list (PEM bundle, a multi-line string).
+
+    If the destination is resolved to an HTTPS URL, a CA list is required
+    to establish a secure connection.
+
+    The CA list can be provided over a relation via the ``ca_relation_name``
+    argument, as a fixed string via the ``ca_data`` argument, or the system CA
+    list will be used if the earlier two are both ``None``.
+    """
+
+    def __init__(
```

Comment:
not sure exactly where we should put the code to do this, but:
as soon as we have a trace ID, would you mind logging it (preferably in a way that's compatible with the existing log message) so we can pick it up with jhack?

https://discourse.charmhub.io/t/how-to-instrument-your-charm-with-tracing-telemetry/11012#p-26285-mapping-events-to-traces-with-jhack-tail-1

Reply:
Deferred, tracked in #1633